livekit-agents/livekit/agents/beta/workflows/task_group.py

```diff
@@ -0,0 +1,128 @@
+
+
+class TaskGroup(AgentTask[TaskGroupResult]):
+    def __init__(
+        self,
+        *,
+        summarize_chat_ctx: bool = True,
+        chat_ctx: NotGivenOr[llm.ChatContext] = NOT_GIVEN,
+    ):
+        """Creates a TaskOrchestrator instance."""
+        super().__init__(instructions="*empty*", chat_ctx=chat_ctx, llm=None)
+
+        self._summarize_chat_ctx = summarize_chat_ctx
+        self._visited_tasks = set()
+        self._registered_factories: OrderedDict[str, _FactoryInfo] = OrderedDict()
+
+    def add(self, task_factory: Callable[[], AgentTask], *, id: str, description: str) -> Self:
+        self._registered_factories[id] = _FactoryInfo(
+            task_factory=task_factory, id=id, description=description
+        )
+        return self
+
+    async def on_enter(self):
+        task_stack = list(self._registered_factories.keys())
+        task_results: dict[str, Any] = {}
+
+        while len(task_stack) > 0:
+            task_id = task_stack.pop(0)
+            factory_info = self._registered_factories[task_id]
+
+            shared_chat_ctx = self.chat_ctx.copy()
+            await self._current_task.update_chat_ctx(shared_chat_ctx)
+            self._current_task = factory_info.task_factory()
+
+            current_tools = self._current_agent_task.tools
+            current_tools.append(self._build_out_of_scope_tool(active_task_id=task_id))
+            await self._current_task.update_tools(current_tools)
+
+            try:
+                self._visited_tasks.add(task_id)
+                res = await self._current_task
+                task_results[id] = res
+            except _OutOfScopeError as e:
+                task_stack.insert(0, task_id)
+                task_stack.insert(0, e.target_task_id)
+                continue
+            except Exception as e:
+                self.complete(e)
+                break
+
+        try:
+            if self._summarize_chat_ctx:
+                # when a task is done, the chat_ctx is going to be merged with the "caller" chat_ctx
+                # enabling summarization will result on only one ChatMessage added.
+                summarized_chat_ctx = await self.chat_ctx.copy(exclude_instructions=True).summarize(
+                    keep_last_turns=0
+                )
+                await self.update_chat_ctx(summarized_chat_ctx)
+        except Exception as e:
+            self.complete(RuntimeError(f"failed to summarize the chat_ctx: {e}"))
+
+        self.complete(TaskGroupResult(task_results=task_results))
+
+    def _build_out_of_scope_tool(self, *, active_task_id: str) -> FunctionTool | None:
+        if not self._visited_tasks:
+            return None
+
+        # Only allow to regress to already visited tasks
+        task_ids = self._visited_tasks.copy()
+        task_ids.discard(active_task_id)
+        task_repr = {f.id: f.description for f in self._registered_factories if f.id in task_ids}
+
+        description = (
```

Comment:
this description seems very specific to the style of agent where it's modifying information in another task.. is it robust enough for other use cases as a task?

should we also allow users to skip around? currently it's only allowing you to go back to previous tasks. I'm wondering if this should be specified on each task itself either:
* can be revisited
* can be skipped ahead

Reply:
@davidzhao Do you have any examples or scenarios where you’d want a task that can be skipped ahead?

> `is it robust enough for other use cases as a task?`

For now we're mostly focused on input collection, do you have other use cases in mind?